### PR TITLE
Intercomparison covermaps

### PR DESCRIPTION
**Pull request by adamy03**

### CODE REVIEW

File: notebooks/intercomparison-covermaps.ipynb [hunk 1]
```diff
@@ -0,0 +1,1346 @@
+{
```
Comment (ivanzvonkov) [inline notebooks/intercomparison-covermaps.ipynb:1]: <p>Is this meant to be filled out?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/crop-mask/pull/228/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (ivanzvonkov) [inline notebooks/intercomparison-covermaps.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    def rasterExtraction(image, resolution, f_collection):
</pre><p>nit: python function names usually use <code>lower_case_underscore</code> https://peps.python.org/pep-0008/#function-and-variable-names </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/crop-mask/pull/228/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (ivanzvonkov) [inline notebooks/intercomparison-covermaps.ipynb:1]: <p>nit: I suggest writing this out with some new lines so it's more legible</p><pre class="ql-syntax" spellcheck="false">new_report = pd.DataFrame(data = {
    "dataset": dataset,
    "accuracy": report["accuracy"],
    ....
})
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/crop-mask/pull/228/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (ivanzvonkov) [inline notebooks/intercomparison-covermaps.ipynb:1]: <ol><li><code>dir</code> is also a python command so I suggest not using it as a variable</li><li><code>Path</code> from the <code>pathlib</code> library, once you get a hang of it, it's a lot cleaner than using strings for paths</li></ol><pre class="ql-syntax" spellcheck="false">from openmapflow.config import PROJECT_ROOT, DataPaths # This uses pathlib behind the scenes

all_dataset_paths = (PROJECT_ROOT / DataPaths.DATASETS).glob("*")
test_paths = [p for p in all_dataset_paths if p.stem in ["Kenya", "Togo"]

...

for p in test_paths:
   df = pd.read_csv(p)
   df["country"] = p.stem
</pre><p> </p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/crop-mask/pull/228/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (ivanzvonkov) [inline notebooks/intercomparison-covermaps.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #17.        df["test_class"] = df["class_probability"].apply(lambda x: 1 if x&gt;=0.5 else 0)
</pre><p>You are counting 0.5 as a crop here is that the intended behavior?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/crop-mask/pull/228/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (ivanzvonkov) [inline notebooks/intercomparison-covermaps.ipynb:1]: <p>In other rows you output the dataframe head but here you do that info? Any particular reason for that?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/crop-mask/pull/228/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (ivanzvonkov) [inline notebooks/intercomparison-covermaps.ipynb:1]: <p>If it's for section 2 why not create it there?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/crop-mask/pull/228/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (ivanzvonkov) [inline notebooks/intercomparison-covermaps.ipynb:1]: <p>Also suggest using pathlib here, is this going to be using Zenodo or Google Cloud long term?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/crop-mask/pull/228/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (ivanzvonkov) [inline notebooks/intercomparison-covermaps.ipynb:1]: <p>Okay so I see you create a new dataframe for each predicted map, merge the test points to the predicted map and calculate the metrics on that dataframe.</p><p>This approach makes sense!</p><p>One small drawback is that it makes comparison between the other maps difficult. For example do Glad and ESA have the same mistakes?</p><p>Instead of creating multiple dataframes I suggest adding to one single dataframe and then computing all metrics on that one single dataframe:</p><p> </p><pre class="ql-syntax" spellcheck="false">esa = ee.ImageCollection("ESA/WorldCover/v100")
esa_results = esa.filterBounds(test_coll).map(lambda x: rasterExtraction(x, 10, test_coll)).flatten()
esa_sampled = geemap.ee_to_gdf(esa_results)
esa_sampled["esa_class"] = esa_sampled["Map"].apply(lambda x: map_values(x, 40))
test_data["esa"] = pd.merge(test_data, esa_sampled, on=["lat", "lon"], how="left")["esa_class"]

glad = ee.ImageCollection("users/potapovpeter/Global_cropland_2019")
glad_results = glad.filterBounds(test_coll).map(lambda x: rasterExtraction(x, 30, test_coll)).flatten()
glad_sampled = geemap.ee_to_gdf(glad_results)
test_data["glad"] = pd.merge(test_data, glad_sampled, on=["lat", "lon"], how="left")["b1"]

# Then metrics only need to be ran once
results = {"Kenya": pd.DataFrame(), "Togo": pd.DataFrame()}
for country, df in test_data.groubpy("country"):
    for dataset in ["esa", "glad"]:
        report = classification_report(df["test_class"], df[dataset])
        results[country] = report_to_row(dataset, report, results[country])
</pre><p> </p><p>Let me know what you think!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/crop-mask/pull/228/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='45'/>

**Review by ivanzvonkov — CHANGES_REQUESTED** [review]
Hello @adamy03, this is great work for an initial round!
I added some comments which are better viewed in ReviewNB (the indentation is there)
I have one major comment about making one dataframe vs many, other comments are all minor formatting things.
This is all feedback so feel free to disagree!

File: notebooks/intercomparison-covermaps.ipynb (hunk 1 above)
Comment (adamy03) [reply notebooks/intercomparison-covermaps.ipynb:1]: <p>This was from some debugging I was doing, but I can display it as head if it looks cleaner.</p>

Comment (adamy03) [reply notebooks/intercomparison-covermaps.ipynb:1]: <p>Yes, this was an assumption that I was making. Not sure if this is the best approach for our purposes - should we threshold another way (i.e [0, 0.5] for non-crop, (0.5, 1] for crop)?</p><p> </p><p>In any case, I will make a note about how thresholding is done in the notebook. </p>

Comment (adamy03) [reply notebooks/intercomparison-covermaps.ipynb:1]: <p>Yes. I intend to replace all the harvest paths with gcloud paths to the wanted files. </p>

**Comment by ivanzvonkov** [comment]
@adamy03 let me know if you have any questions about any of my comments about the notebook

File: notebooks/intercomparison-covermaps.ipynb [hunk 2]
```diff
@@ -0,0 +1,1881 @@
...
```
Comment (ivanzvonkov) [inline notebooks/intercomparison-covermaps.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    cop_sampled["cop_class"] = cop_sampled["discrete_classification"].apply(lambda x: map_values(x, 40))
</pre><p> </p><p>nit: a one liner instead of map_values here is probably easier to understand</p><p><code>lambda x: 1 if x == 40 else 0</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/crop-mask/pull/228/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='36'/>

Comment (ivanzvonkov) [inline notebooks/intercomparison-covermaps.ipynb:1]: <p>Nice this looks pretty good!</p><p>I personally prefer flattening control flows in for loops by using <code>continue</code> but you don't necessarily have to take this feedback. Here's how it could look with <code>continue</code> </p><pre class="ql-syntax" spellcheck="false">for dataset in ["cop", "esa", "glad", "harvest_togo", "harvest_kenya", "harvest_tanzania"]:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; if pd.isnull(df[dataset]).all():
            print(f"{dataset} is empty"
            continue
        if len(df[dataset].unique()) &gt; 1:
            print(f"{dataset}: found multiple datasets.")
            continue

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; print(country + ": " + dataset)&nbsp;&nbsp;&nbsp; 
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;clean_df = df[["test_class", dataset]].dropna() # Remove na values
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;if len(clean_df) &lt; 10:
            print(f"{dataset}" has less than 10 examples")
            continue

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;report = classification_report(clean_df["test_class"], temp[dataset], output_dict= True)&nbsp;&nbsp;&nbsp; &nbsp;

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; results[country] = report_to_row(dataset, report, results[country])

</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nasaharvest/crop-mask/pull/228/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='49'/>

Comment (ivanzvonkov) [reply notebooks/intercomparison-covermaps.ipynb:1]: Oh no this did not come out as nicely outside ReviewNB

File: src/compare_covermaps.py [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+import cartopy.io.shapereader as shpreader
+import ee
+import geemap
+import geopandas as gpd
+import pandas as pd
+from shapely.geometry import GeometryCollection
+from sklearn.metrics import classification_report
+
+# ee.Authenticate()
+ee.Initialize()
```
Comment (ivanzvonkov) [inline src/compare_covermaps.py:13]: I suggest removing both of these lines and let the user run this command from the notebook. That way it does not execute every time you import the file

**Comment by ivanzvonkov** [comment]
Could you address some of the pre-commit linting failures which are visible if you click "Details": https://results.pre-commit.ci/run/github/324755397/1678902480.NNQY7OIuRS2amcjO41mB8A

image

**Review by ivanzvonkov — APPROVED** [review]
On the whole this looks good, It can be merged once the `pre-commit` checks are addressed (see above comment). Normally
